### PR TITLE
Merge master downstream and update hie-bios

### LINKED ISSUE #9: Race between closing a Virtual File and persisting a Virtual File
We observed in #8 a race condition that we could reproduce in the following way:
Open a file and as soon as one can observe compilation progress, close the file in the editor.
This reliably would cause an error in `haskell-lsp`:`src/Language/Haskell/LSP/VFS.hs:131:16`.

To see what happens in hie, I created this sequence diagram:
Basic Sequence Diagram

In words: when we receive a `NotDidOpenDocument`, we send an document update to the dispatcher and then to the `GHC Thread`. In this update we persist the Virtual File that has been freshly created to a temporary file in the file system.
I assume we send this message to the ghc thread because previously, ghc-mod took care of Virtual Files, and ghcmod lived in the ghc thread. 

However, when we send a message to the ghc thread, the action is run in a context. This context is the action: `getCradle uri (\lc -> loadCradle df lc >> action)`. So, now before we actually try to persist the Virtual File to the file system, we load the cradle. After the cradle has been loaded, we persist the Virtual File.
In the meantime, if a `NotDidCloseDocument` is processed, it will remove the Virtual File from the VFS. The problem is now, that `NotDidCloseDocument` does not send a message to the ghc thread and can immediately remove the Virtual File, while the processing of the `NotDidOpenDocument` action is still stalled on the `loadCradle` function.

A work around is presented here:
Basic Sequence Diagram(2)

In short, we persist the Virtual File immediately in a blocking action, so no race can occur.

We can do this, because we can persist the Virtual File immediately, there is no ghc-mod that manages that for us.
How did that work before? Since VFS and persisting the Virtual File to the file system were two distinct actions, taken care of by different actors that knew nothing about each other. Closing a virtual file had no impact on ghcmod persisting that file in the file system.

### LINKED ISSUE #11: Hie crashes if cabal file has a syntax error
Can be reproduced by trying to use HIE on a project with a wrong cabal file.
Example output:

<details>
<summary>Crash on unsat dependencies</summary>

```
2019-09-29 16:07:47.846291691 [ThreadId 4] - Using stack GHC version
2019-09-29 16:07:49.891036604 [ThreadId 4] - Run entered for HIE(hie-8.6.4) Version 1.0.0.0 x86_64 ghc-8.6.4
2019-09-29 16:07:49.891179924 [ThreadId 4] - Current directory:/home/baldr/Documents/definition
2019-09-29 16:07:49.896248263 [ThreadId 9] - Using stack GHC version
Using hie version: Version 1.0.0.0 x86_64 ghc-8.6.4
("lookupCradle","/home/baldr/Documents/definition/Bar.hs",GhcModuleCache {cradleCache = Data.Trie.fromList [], uriCaches = fromList [], currentCradle = Nothing})
("New cradle","/home/baldr/Documents/definition/Bar.hs")
Cradle {cradleRootDir = "/home/baldr/Documents/definition", cradleOptsProg = CradleAction: stack}
Using hoogle db at: /home/baldr/.hoogle/default-haskell-5.0.17.hoo
hie-8.6.4: (ExitFailure 1,"Unable to parse cabal file /home/baldr/Documents/definition/definitions.cabal\n\n- 9:33: \nunexpected \"b\"\nexpecting space, opening paren or operator\n\nbase < 5 && base >= 5\n\n\n","")
CallStack (from HasCallStack):
  error, called at src/HIE/Bios/Cradle.hs:299:18 in hie-bios-0.2.1-1ae2600e88e53435a0e65bf4e00ef6b6bfbfe9f4d9259323955b4fc3b7c62d16:HIE.Bios.Cradle
```
</details>

Expected behaviour: report the problem rather than just crashing.

### PR DESCRIPTION
**Pull request by fendor**
Resolves merge conflicts.
Fixes build on all supported ghc versions.
Fixes all warnings in the code base.
Reduce debugging noise.
closes #9. 
works around #10. 
kind of fixes #11. HIE does not crash, but no proper diagnostics are produced.

### CODE REVIEW

**Comment by mpickering** [comment]
Did you test much after doing the merge?

**Comment by fendor** [comment]
I am currently constantly dogfooding and mainly observed the problem we already discussed regarding the virtual file system.
I am pretty sure, I found the problem but not how to properly fix it.
closing a document removes it from the vfs, but in the type checking process persistVirtualFile is manually called which throws an error if the document had been closed.
I want to add a test for that before merge.
Also, I want to know if all uses of persistVirtualFile are prone to this race. Need to investigate it a bit more. 

I also observed an error that Data.Text.last was executed on an empty string. I am not sure if that was already present on master and it happened only once.

Other than that, it was pretty good, and the ide felt more responsive.

**Comment by mpickering** [comment]
I discussed this race condition problem with @fendor on IRC. We concluded that it was difficult to properly fix the race but could guard against it by more carefully using `persistVirtualFile`.

I'll give you merge access @fendor so you can merge it when you're happy.

**Comment by mpickering** [comment]
Ready to merge?

**Comment by fendor** [comment]
No, there are still two issues left, I intend to work on before the merge.

File: cabal.project [hunk 1]
```diff
@@ -15,7 +13,7 @@ package haskell-ide-engine
 
 allow-newer: floskell:all
 
-profiling: false
+profiling: true
```
Comment (fendor) [inline cabal.project:17]: need to undo this

**Pull request merged**